### PR TITLE
Refactor: change response status codes for invalid inputs (409 to 400) in register endpoint

### LINKED ISSUE #653: Bug: HTTPStatus CONFLICT is returned instead of BAD_REQUEST for invalid input POST /register
**Describe the bug**

HTTPStatus.CONFLICT is returned for invalid input instead of HTTPStatus.BAD_REQUEST

**To Reproduce**

1. Run `python run.py` command
2. Go to Swagger UI and try to register new user with invalid input
3. You will see error message with 409 CONFLICT instead of BAD_REQUEST 400 being returned.
You can also see the expected response on Swagger UI that have invalid input returned as HTTPStatus.CONFLICT (409)

**Expected behavior**
The error return on invalid inputs need to be HTTPStatus.BAD_REQUEST (400) instead of CONFLICT  (409)

**Screenshots**
Screen Shot 2020-06-28 at 2 03 47 pm

**Desktop (please complete the following information):**
 - OS: MacOS
 - Browser: Chrome
 - Version ??

**Additional context**
This is caused by the recent [PR#650](https://github.com/anitab-org/mentorship-backend/pull/650) that got merged to develop.

### PR DESCRIPTION
**Pull request by NenadPantelic**
### Description

Change response status code for the scenario when the user provides invalid input in register action (shorter username/password, space in the username....). Status codes for this situation are changed to 400 (`BAD_REQUEST`). 
Fixes #653

### Type of Change:
- Code

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)

### How Has This Been Tested?

Tested with:
1. Swagger UI 
image
image
<hr>
Executed twice:

image
image

2. with unit-tests.  Added to [resources test](https://github.com/NenadPantelic/mentorship-backend/blob/653-fix-response-status-codes-in-register/tests/users/test_api_resources.py) - methods `test_user_registration_with_invalid_inputs_bad_request` and `test_user_registration_with_invalid_inputs_conflict_username`

### Checklist:

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas

**Code/Quality Assurance Only**

- [x] My changes generate no new warnings 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by NenadPantelic** [comment]
@vj-codes @mtreacy002 Can you please review this PR? Thanks in advance :slightly_smiling_face:

**Comment by vj-codes** [comment]
@NenadPantelic hey congratulations on your first PR to the repo 🎉 
Can you please test the remaining 3 cases under HTTP.Status BAD_REQUEST and update in comments if it's working as expected? Looking at the code it's bound to work but just to be 100% sure and then PR can be approved:)

**Comment by NenadPantelic** [comment]
> @NenadPantelic hey congratulations on your first PR to the repo 
> Can you please test the remaining 3 cases under HTTP.Status BAD_REQUEST and update in comments if it's working as expected? Looking at the code it's bound to work but just to be 100% sure and then PR can be approved:)

Hi, @vj-codes, thanks for the review.  I suppose you think of testing other input scenarios:
1. username is empty
2. space in the password
3.  invalid email

Is that right?

**Comment by vj-codes** [comment]
@NenadPantelic yes absolutely right and please squash your commits .

**Comment by NenadPantelic** [comment]
@mtreacy002 @vj-codes Can you review it now? I added separated tests for all use-case scenarios when the user provides invalid input :slightly_smiling_face:

**Comment by vj-codes** [comment]
@devkapilbansal can you review this ? Your name is not coming in Review Requests section

**Review by vj-codes — DISMISSED** [review]
Congrats on your first contribution to the repo @NenadPantelic 🎉
LGTM !

**Comment by devkapilbansal** [comment]
> @devkapilbansal can you review this ? Your name is not coming in Review Requests section

Will check by tomorrow :+1:

File: tests/users/test_api_resources.py [hunk 1]
```diff
@@ -137,6 +135,161 @@ def test_user_registration_mentee(self, send_email_function):
...
+                        terms_and_conditions_checked=user1[
+                            "terms_and_conditions_checked"
+                        ],
+                    )
+                ),
+                follow_redirects=True,
+                content_type="application/json",
+            )
+
+            user = UserModel.query.filter_by(username=user1["username"]).first()
```
Comment (devkapilbansal) [inline tests/users/test_api_resources.py:227]: Hi @NenadPantelic can you please assert the message returned by response too? Same for other tests too

**Review by devkapilbansal — CHANGES_REQUESTED** [review]

File: tests/users/test_api_resources.py (hunk 1 above)
Comment (NenadPantelic) [reply tests/users/test_api_resources.py:227]: Yeah @devkapilbansal, that sounds reasonable :wink:

**Comment by NenadPantelic** [comment]
@vj-codes @devkapilbansal I added response message comparison. PR is again ready for the review :repeat:

**Comment by devkapilbansal** [comment]
Thanks @NenadPantelic 👍

File: app/messages.py [hunk 2]
```diff
@@ -54,8 +54,8 @@
 AUTHORISATION_TOKEN_IS_MISSING = {"message": "The authorization token is" " missing!"}
 DESCRIPTION_FIELD_IS_MISSING = {"message": "Description field is missing."}
 COMMENT_FIELD_IS_MISSING = {"message": "Comment field is missing."}
-USERNAME_FIELD_IS_EMPTY = {
-    "message": f"The username field has to be longer than {USERNAME_MIN_LENGTH} characters and shorter than {USERNAME_MAX_LENGTH} characters."
+USERNAME_HAS_INVALID_LENGTH = {
+    "message": f"The username field has to be longer than {USERNAME_MIN_LENGTH - 1} characters and shorter than {USERNAME_MAX_LENGTH + 1} characters."
```
Comment (devkapilbansal) [inline app/messages.py:58]: Hi, @NenadPantelic why you use `-1` and `+1` here?

**Comment by NenadPantelic** [comment]
> Thanks @NenadPantelic
@devkapilbansal 
That's because of  the validations methods. There is mismatch between messages length and validation methods that check if the username and password satisfies length constraint. E.g. if you provide username that has less than 5 chars -> `test` on of the validation methods will throw the exception:
```
 "The {field_name} field has to be longer than {min_limit} "
            "characters and shorter than {max_limit} characters.".format(
                field_name=field_name,
                min_limit=min_length - 1,
                max_limit=max_length + 1,
            )
```
So, the error message that is planned to be thrown is not the same as the one which is actually returned. In that case, the test expects to get: `The username field has to be longer than 4 characters and shorter than 26 characters.` and the validation method throws: `The username field has to be longer than 5 characters and shorter than 25 characters.`. To solve that mismatch, I had to make them the same. 

I think that the best solution was to change those validation methods so that they simply return predefined messages without creating their own response messages. I didn't do that because that's not the purpose of this issue. I would create a new issue -> refactor validation methods to follow error response messages. 
Those messages should be static and validation messages should just validate input length and throw them as they're, without creating new messages. I'm not sure If I explained this well enough or just leveled up the confusion :sweat_smile:

**Review by devkapilbansal — APPROVED** [review]
The changes made in this PR were tested locally. Following are the results:

1. Code review - Done

2. All possible responses (positive and negative tests) were tested as below:

  * _Test1 Description_  Change response status code in registration endpoint
    _Screenshot/gif_:  


    _Expected Result_:  Conflict should be returned only when the user with those details already exist.
    _Actual Result_: Same as expected
    
3. Additional testcases covered: All tests needed are inlcuded in tests directory and passed as expected

4. Status of PR Changed to: Ready to Merge.

**Comment by devkapilbansal** [comment]
@vj-codes @isabelcosta review this pull request

**Comment by devkapilbansal** [comment]
> > Thanks @NenadPantelic
> > @devkapilbansal
> > That's because of  the validations methods. There is mismatch between messages length and validation methods that check if the username and password satisfies length constraint. E.g. if you provide username that has less than 5 chars -> `test` on of the validation methods will throw the exception:
> 
> ```
>  "The {field_name} field has to be longer than {min_limit} "
>             "characters and shorter than {max_limit} characters.".format(
>                 field_name=field_name,
>                 min_limit=min_length - 1,
>                 max_limit=max_length + 1,
>             )
> ```
> 
> So, the error message that is planned to be thrown is not the same as the one which is actually returned. In that case, the test expects to get: `The username field has to be longer than 4 characters and shorter than 26 characters.` and the validation method throws: `The username field has to be longer than 5 characters and shorter than 25 characters.`. To solve that mismatch, I had to make them the same.
> 
> I think that the best solution was to change those validation methods so that they simply return predefined messages without creating their own response messages. I didn't do that because that's not the purpose of this issue. I would create a new issue -> refactor validation methods to follow error response messages.
> Those messages should be static and validation messages should just validate input length and throw them as they're, without creating new messages. I'm not sure If I explained this well enough or just leveled up the confusion

Actually I understand that and that's why delete the comment too but thanks for the explanation. :clap:

File: tests/users/test_api_resources.py (hunk 1 above)
Comment (devkapilbansal) [reply tests/users/test_api_resources.py:227]: Thanks for considering

**Comment by NenadPantelic** [comment]
> Actually I understand that and that's why delete the comment too but thanks for the explanation.


@devkapilbansal 
Yeah, I saw that, but I have typed it already and was sorry to delete it :joy:. I can open the refactoring issue for that. It will be cleaner code in that way (to recode those validation methods). Thanks for the review :handshake:

**Review by vj-codes — APPROVED** [review]

**Comment by vj-codes** [comment]
Adding label ready to merge since testing is done above 🚀

**Review by isabelcosta — APPROVED** [review]
Awesome work @NenadPantelic :) Thank you for pinging me on Zulip.
Great that you added tests for this 👏 🎉

**Comment by NenadPantelic** [comment]
> Awesome work @NenadPantelic :) Thank you for pinging me on Zulip.
> Great that you added tests for this

Thanks @isabelcosta :blush:

**Comment by mtreacy002** [comment]
Thanks for the fix, @NenadPantelic 🎉

**Comment by NenadPantelic** [comment]
> Thanks for the fix, @NenadPantelic

You're welcome @mtreacy002 :blush:

**Pull request merged**